integration/simulation/network/dockerenclave.go

```diff
@@ -99,16 +180,88 @@ func (n *basicNetworkOfNodesWithDockerEnclave) Create(params *params.SimParams,
 }
 
 func (n *basicNetworkOfNodesWithDockerEnclave) TearDown() {
+	n.gethNetwork.StopNodes()
 	for _, client := range n.obscuroClients {
 		temp := client
 		go func() {
 			defer (*temp).Stop()
 			(*temp).Call(nil, obscuroclient.RPCStopHost) //nolint:errcheck
 		}()
 	}
+	terminateDockerContainers(n.ctx, n.client, n.containerIDs)
+}
+
+// Checks the required Docker images exist.
+func dockerImagesAvailable(ctx context.Context, cli *client.Client) bool {
+	images, _ := cli.ImageList(ctx, types.ImageListOptions{})
+	for _, image := range images {
+		for _, tag := range image.RepoTags {
+			if strings.Contains(tag, enclaveDockerImg) {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// Creates the test Docker containers.
+func createDockerContainers(ctx context.Context, client *client.Client, numOfNodes int, startPort int, mngmtCtrAddr string, erc20Addr string) map[string]string {
+	var enclavePorts []string
+	for i := 0; i < numOfNodes; i++ {
+		// We assign an enclave port to each enclave service on the network.
+		enclavePorts = append(enclavePorts, fmt.Sprintf("%d", startPort+DefaultEnclaveOffset+i))
+	}
+
+	containerIDs := map[string]string{}
+	for idx, port := range enclavePorts {
+		nodeID := common.BigToAddress(big.NewInt(int64(idx + 1))).Hex()
+		containerConfig := &container.Config{
+			Image: enclaveDockerImg,
+			Cmd: []string{
+				"--" + enclaverunner.HostIDName, nodeID,
+				"--" + enclaverunner.AddressName, enclaveAddress,
+				"--" + enclaverunner.ManagementContractAddressName, mngmtCtrAddr,
+				"--" + enclaverunner.Erc20contractAddrsName, erc20Addr,
+			},
+		}
+		hostConfig := &container.HostConfig{
+			PortBindings: nat.PortMap{nat.Port(enclaveDockerPort): []nat.PortBinding{{HostIP: Localhost, HostPort: port}}},
+		}
+
+		resp, err := client.ContainerCreate(ctx, containerConfig, hostConfig, nil, "")
+		if err != nil {
+			panic(err)
+		}
+		containerIDs[resp.ID] = port
+	}
+
+	return containerIDs
+}
+
+// Stops and removes the test Docker containers.
+func terminateDockerContainers(ctx context.Context, cli *client.Client, containerIDs map[string]string) {
+	for id := range containerIDs {
+		// timeout := -time.Nanosecond // A negative timeout means forceful termination.
+		err1 := cli.ContainerStop(ctx, id, nil)
+		if err1 != nil {
+			fmt.Printf("Could not stop the container %v : %s\n", id, err1)
+			continue
+		}
+
+		err2 := cli.ContainerRemove(ctx, id, types.ContainerRemoveOptions{
+			RemoveVolumes: true,
+			RemoveLinks:   false,
+			Force:         true,
+		})
+		if err2 != nil {
+			fmt.Printf("Could not remove the container %v : %s\n", id, err2)
+			continue
+		}
+
+		fmt.Printf("Stopped and removed container %v\n", id)
+	}
 
-	for _, node := range n.ethNodes {
-		temp := node
-		go temp.Stop()
+	if err := cli.Close(); err != nil {
+		panic(err)
```

Comment:
@tudor-malene If we panic here, we may interrupt the shutdown of the Geth network and clients that is happening on separate goroutines.

Reply:
Ah, I see what you mean now. I'll have another look at this stuff then